Partition SciJava Ops into separate modules

gselzer opened the pull request:
This PR partitions SciJava Ops into a set of new JPMS/Maven modules:
* SciJava Ops API: provides a set of external API for use by third party libraries who **only** wish to use SciJava Ops / ImageJ Ops2
* SciJava Ops SPI: provides a set of annotations/service interfaces for use by third party libraries who wish to **extend** SciJava Ops / ImageJ Ops2 with **their own implementations**
* SciJava Ops Engine: provides a default Op matching implementation, with Simplification, Adaptation, etc.
* SciJava Ops Discovery: provides an interface for class discovery. Currently, SciJava Ops Engine implements this with Plugin Discovery, but we could easily add a Spring Boot implementation

This allows much more control over which sections of the library will be exported, and will result in cleaner code. This will also narrow our dependency on various libraries to certain submodules of SciJava Ops, making it easier to migrate the library to new implementations of, say, dependency injection or implementation discovery.

TODO:
* [ ] Where should we store the API tests (for example, `OpBuilderTest`)? Ideally, we'd keep them within SciJava Ops API, however we don't have the implementations required to thoroughly test, for example, `OpBuilder` within SciJava Ops API. We could always keep them within SciJava Ops Engine, but this seems like bad practice. \
* [ ] Better partition the SciJava Common dependency. This is hard, because all Ops (both test and main Ops) currently have the dependency on `Plugin`.
* [ ] Merge #28, #31, #39

Closes scijava/scijava#63, scijava/scijava#61

gselzer committed:
therapi-runtime-javadoc: first cut

gselzer committed:
Scrape javadoc for parameter data

gselzer committed:
Member: add description field

This will allow us to retain the descriptions scraped from the javadoc

gselzer committed:
Test obtaining the descriptions of parameters

gselzer committed:
Convert Parameter annotations to javadoc

Accomplished via this script:

  perl -0777 -i -pe 's/(\@Plugin[^\n]*\n)\@Parameter.key = "([^"]*)".\n/\1\/**\n * TODO\n *\n * \@param \2\n/igs' **/*.java
  perl -0777 -i -pe 's/\@Parameter.key = "([^"]*)".\n/ * \@param \1\n/igs' **/*.java
  perl -0777 -i -pe 's/( \* \@param[^\n]*\n)([^ ])/\1 *\/\n\2/igs' **/*.java

Plus vim macros and manual tidying.

gselzer committed:
Avoid searching for @Parameter annotations

gselzer committed:
Remove obsolete Parameter imports

gselzer committed:
Move @Override annotations around javadoc

gselzer committed:
Remove unused ItemIO imports

These imports were used by the @Parameter declarations; since we removed
those declarations we can also remove the imports

gselzer committed:
Add descriptions to OpString

gselzer committed:
Update paramString for Member descriptions

gselzer committed:
Clean TODOs, output names

gselzer committed:
Consolidate OpInfo printout methods

There were two different methods that wrote printouts for OpInfos, using
rather different syntaxes. We decided that one syntax was better, and
added member highlight support (which was only available using the other
    syntax).

gselzer committed:
Remove remove-params.sh script

It has outlived its usefulness to us. RIP

ctrueden committed:
Implement taglets for OpField-related parameters

gselzer committed:
Update parent to pom-scijava 30.0.0

gselzer committed:
Make @input and @param identical within Javadoc

This change applies only to method and class Ops, however it enables
@input synonymous with @param and @output synonymous with @return. This
also allows us to use @container and @mutable on Ops written as classes
and as methods. This then changes the paradigm to make the parsing of
@param and @return a bonus, but favors using @input and @output
respectively.

gselzer committed:
Prevent NSEExceptions when OpMethod type incorrect

Suppose that an Inplace OpMethod is annotated with the functional
interface BiFunction. We must catch these situations quickly and fail
before the struct is generated, as it is not the job of ParameterStructs
to deal with incorrectly annotated methods

gselzer committed:
Make javadoc parsing robust against OpDependency

gselzer committed:
Comment method

gselzer committed:
Track OpDependencies with a taglet

Used so that we maintain one tag per input for Ops written as methods.
This taglet should not appear on Ops written as Fields or as Classes
(although it will not break anything if they do, the javadoc parser just
 ignores these lines)

gselzer committed:
Add javadoc scraping tests

gselzer committed:
Comment JavadocParameterData

gselzer committed:
Delete javax.annotation.processing.Processor file

Unnecessary - maven handles this for us

gselzer committed:
generate module-info via Apache Velocity

gselzer committed:
Remove Parameter/ItemIO imports from .vm files

Not sure how these escaped the purge

gselzer committed:
Establish platform independence in Module testing

gselzer committed:
pom: remove dependency on therapi-runtime-javadoc-scribe

The annotation processor side of therapi should not be transitively
passed to projects depending on scijava ops - we only need the side of
therapi that looks up the javadoc

gselzer committed:
Disable custom javadoc taglets for methods/classes

This ensures that everyone will be writing proper javadoc whenever
proper javadoc will be sufficient for our purposes

gselzer committed:
Remove explicit notation of annotation processors

gselzer committed:
scijava/scijava-ops/pom.xml: Remove versioning

Rely on pom-scijava for this

gselzer committed:
Remove annotation processor options

They are not being used by our annotation processors, so there is no
reason to keep them.

gselzer committed:
Move therapi annotation processor upstream

Doing this allows us to remove the maven compiler plugin conifguration
from scijava-ops' POM

gselzer committed:
scijava-ops/pom.xml: Delete jar-plugin config

This is not needed; pom-scijava-base does this for us

gselzer committed:
Remove compiler config

When we move the module-path annotation processors upstream, we want to
make as few changes as possible. These options are not necessary, thus
we'll remove them

gselzer committed:
Change switch case to input from param

gselzer committed:
Further flesh out JavadocParameterTest

gselzer committed:
JavadocParameterTest: remove Parameter annotations

gselzer committed:
Initialize scijava-struct module

gselzer committed:
scijava-ops: depend on scijava-struct

This commit removes the org.scijava.struct package, since it was moved
to its own module. scijava-ops now depends on scijava-struct

ctrueden committed:
Install scijava-taglets before the reactor build

It's a dependency of the maven-javadoc-plugin, which means it needs
to be installed before generating javadocs for any of the modules.

gselzer committed:
WIPPY: add package cycle enforcer to incubator

gselzer committed:
Move Op to org.scijava.ops

gselzer committed:
Move OpCollection to org.scijava.ops

gselzer committed:
Move OpBuilder to org.scijava.ops

OpEnvironment returns an OpBuilder, so OpBuilder should be in the same
package

gselzer committed:
Move Hints/OpHints to org.scijava.ops

Since OpEnvironments have Hints, the Hints need to be in the base
package.

gselzer committed:
Extract OpService interface

gselzer committed:
Extract OpRef interface

This commit also makes some changes to SimplifiedOpRef that make it more
correct. I chose to throw UnsupportedOperationException for many of
the standard OpRef methods because the concrete type that the
SimplifiedOpRef is asking for is rather indeterminate - it is really
asking for an Op that is of a set of Types, so to pin a particualar set
of arguments, a particular output type, etc. is not possible until you
have a SimplifiedOpInfo to narrow down these Types. Similarly, we throw
an UnsupportedOperationException for typesMatch because the typing of
the SimplifiedOpRef is unclear. We could match against all possible
permutations of the input/output types, but we don't really have enough
information about how to form the opType. We could probably find a way
to devise all opType permutations, but this would be a pretty
computationally expensive method to run, so for now we'll just throw an
error

gselzer committed:
OpMatchingException: Extend RuntimeException

This prevents us from having to declare it. There was no reason why we
should

gselzer committed:
Move OpCandidate to org.scijava.ops

gselzer committed:
Move util tests from org.scijava.ops to util package

gselzer committed:
Move OpWrapper to org.scijava.ops package

gselzer committed:
Remove HashMaps from FifaceUtils classes

gselzer committed:
WIP: Move fiface utility methods to OpBuilder

gselzer committed:
Remove match methods from fiface util classes

gselzer committed:
Export MatchingUtils methods to scijava-types

These methods do not depend on Ops, so we should export them for broader
usage

gselzer committed:
Remove Parameter annotation

This allows us to later remove the org.scijava.param fork

gselzer committed:
WIP: Switch to MemberParser paradigm

gselzer committed:
remove org.scijava.param package

gselzer committed:
Move ValidityException to org.scijava.ops

This commit also adds an isRequired() field to Member.

gselzer committed:
Extract matcher impl details to impl subpackage

gselzer committed:
Prevent cycle: org.scijava.ops.simplify/struct

gselzer committed:
Prevent cycle: org.scijava.ops.util/org.scijava.ops.struct

gselzer committed:
WIP: Fix ExceptionUtils

gselzer committed:
Resolve package cycle: geom tests

gselzer committed:
Fix package cycle: Persist tests

gselzer committed:
WIP: Use scijava-packages-plugin

NOTE: This commit forward requires the usage of scijava-packages-plugin
(https://github.com/scijava/scijava-packages-plugin). Since no release
has yet been made for this plugin, you must download and build this
repository to use the plugin (and thus build from this commit forward)

gselzer committed:
Add subpackage depedency prevention rule

gselzer committed:
Consolidate create packages

The current situation is untenable since Creators depends on the
net.imagej.ops2.create.kernel and net.imagej.ops2.create.img packages.
Ideally, we'd split out the image- and kernel create Ops into their own
packages like we had, but the kernel and image creator Op fields rely on
the (non-static) Op fields in the main package. Therefore if we moved
these Ops to the subpackages, we would have to duplicate code. To
prevent this duplication, we shall just consolidate to one package.

gselzer committed:
Absorb zernike helper into zernike package

gselzer committed:
Absorb net.imagej.ops2.geom.geom3d.mesh

There seemed to be no reason for this package. The contents did not
really deal with Meshes, and many Ops in the net.imagej.ops2.geom.geom3d
package (that actually imported Mesh) depended on this subpackage (which
is bad). For these reasons, let's just absorb the mesh subpackage into
the geom3d package

gselzer committed:
Separate out API module to contain annotations

gselzer committed:
Remove subpackage dependence rule

We need a TEMPORARY removal of this rule to partition out the modules

gselzer committed:
Refactor ops to use org.scijava.ops.api.Op

gselzer committed:
Use org.scijava.ops.api.OpCollection

gselzer committed:
Use org.scijava.ops.api.OpDependency

gselzer committed:
Use org.scijava.ops.api.OpMethod

gselzer committed:
Use org.scijava.ops.api.OpField

gselzer committed:
Switch to OpInfoGenerator paradigm

This approach has two benefits:
1) Removes the dependency on scijava-common within DefaultOpEnvironment
2) Enables IoC

(Of course we are then just shifting the scijava-common dependency to
 the relevant OpInfoGenerators/DefaultOpService, but this is easier to
 reconcile than a dependency deeper within the OpEnvironment/OpMatcher)

gselzer committed:
Rename scijava-ops -> scijava-ops-engine

gselzer committed:
Rename scijava-ops-api -> scijava-ops-spi

This is much closer to the actual spirit of the project. We are
declaring a set of service annotations/classes that the user may wish to
declare

gselzer committed:
WIP: Partition scijava-ops-api

gselzer committed:
BIG WIP: Add scijava-common dependency

Ideally, we wouldn't need SJC for this, but we need ValidityProblem

gselzer committed:
Scijava-ops-api

gselzer committed:
Re-enforce the prevention of subpackage dependence

gselzer committed:
Move DependencyMember impls to engine.struct package

gselzer committed:
Remove unnecessary exports

gselzer committed:
Partition util package

Since we are exporting this package, we only want the classes designed
to be exported in this package. This spurs the creation of a
util.internal package

gselzer committed:
Create conversionLoss.impl package for impls

Now we only export the interfaces

gselzer committed:
Create package-info for org.scijava.ops.engine

ctrueden commented:
> Where should we store the API tests

I don't know that there's really a such thing as an "API test"—only tests of implementations.

That said, now that Java interfaces can have default implementations, it can get tricky to test those default methods in the API component while testing implementation(s) of abstract methods in the components that implement them. My vote would be to just keep all the tests in the engine component, which is where the implementation gets completed.

gselzer commented:
> > Where should we store the API tests
> 
> I don't know that there's really a such thing as an "API test"—only tests of implementations.
> 
> That said, now that Java interfaces can have default implementations, it can get tricky to test those default methods in the API component while testing implementation(s) of abstract methods in the components that implement them. My vote would be to just keep all the tests in the engine component, which is where the implementation gets completed.

Yeah, it just sucks that we have to have the `OpBuilder` implementation in API. We *could* make an interface for `OpBuilder`...

gselzer committed:
Resolve provenance package cycle

This was a tough decision, as I'd like to keep the provenance stuff
separate from the matcher, but we have to take the OpHistory in the
OpWrappers, so we have to make this change. It is probably worth
considering making a scijava-ops-provenance module that can stand alone
from scijava-ops(-engine)

gselzer commented:
This would also close scijava/scijava#59
